src/main/java/frc/robot/subsystems/LightStrip.java

```diff
@@ -2,41 +2,58 @@
 
 import edu.wpi.first.wpilibj.motorcontrol.Spark;
 import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.constants.Constants;
 import frc.robot.utils.BlinkinPattern;
+import frc.robot.utils.smartshuffleboard.SmartShuffleboard;
 
 import java.util.HashMap;
 import java.util.Map;
-import java.util.concurrent.Executors;
-import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.BooleanSupplier;
 
 public class LightStrip extends SubsystemBase {
-    private static final String baseLogName = "/robot/lightstrip/";
     private final Spark colorSensorPort;
-    private BlinkinPattern pattern = BlinkinPattern.BLACK;
+    private final AtomicReference<BlinkinPattern> pattern = new AtomicReference<>(BlinkinPattern.BLACK);
     private final Map<BooleanSupplier, BlinkinPattern> predicateLightEvents = new HashMap<>();
+    private final AtomicBoolean running = new AtomicBoolean(true);
     public LightStrip(int port) {
         this.colorSensorPort = new Spark(port);
-        Executors.newScheduledThreadPool(1).scheduleAtFixedRate(() -> predicateLightEvents.keySet()
-                .stream()
-                .filter(BooleanSupplier::getAsBoolean)
-                .findFirst()
-                .ifPresent(c -> {
-                            setPattern(predicateLightEvents.get(c));
-                            predicateLightEvents.remove(c);
-                        }
-                ),0,40, TimeUnit.MILLISECONDS);
+        new Thread(() -> {
+            while (running.get()){
+                for (BooleanSupplier supplier : predicateLightEvents.keySet()){
+                    if (supplier.getAsBoolean()){
+                        setPattern(predicateLightEvents.get(supplier));
+                        predicateLightEvents.remove(supplier);
+                        break;
+                    }
+                }
+                try {
+                    Thread.sleep(40);
+                } catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+        }).start();
     }
 
-    public synchronized void setPattern(BlinkinPattern pattern) {
-        this.pattern = pattern;
+    public void setPattern(BlinkinPattern pattern) {
+        this.pattern.set(pattern);
```

Comment:
Seems like there is no real need to protect the `pattern` and the `running` since they are primitives (boolean and a reference), though using an atomic reference here does not hurt. The one thing that does require protecting as it is modified from both threads is the map, which is currently vulnerable.

Reply:
changed to concurrent hashmap